### PR TITLE
Add GPU support

### LINKED ISSUE #66: Re-add the required gpu support
* control over floating point type (this is in already!)
* add HOSTDEVICE to sum functions
* add HOST to update
* add DEVICE to update_device

### PR DESCRIPTION
**Pull request by richfitz**
This PR activates new GPU support in odin.dust.

Changes here are fairly minor:

* split generating gpu-like code from compiling with NVCC
* annotate sum functions with HOSTDEVICE
* annotate device update function with DEVICE
* special treatment of `%%`, `min` and `max`, as found empirically with sircovid

I've also added the sirs example used to develop dustgpu, but we might swap that out for something a bit easier to push around at some point

Requires https://github.com/mrc-ide/dust/pull/187 to be merged, and then the DESCRIPTION here updated to remove the branch pin

The codecov error is it complaining about a file that no longer exists :man_shrugging: 

Fixes #66

### CODE REVIEW

File: DESCRIPTION [hunk 1]
```diff
@@ -36,5 +36,5 @@ Suggests:
 RoxygenNote: 7.1.1
 VignetteBuilder: knitr
 Remotes:
-    mrc-ide/dust,
+    mrc-ide/dust@i73-gpu,
```
Comment (johnlees) [inline DESCRIPTION:39]:
```suggestion
    mrc-ide/dust,
```

**Review by johnlees — APPROVED** [review]
Are there any other odin functions other than these sums that might need a `HOSTDEVICE`?

**Comment by richfitz** [comment]
> Are there any other odin functions other than these sums that might need a `HOSTDEVICE`?

Good question - I just checked and all the other functions are used just for the interface layer (so definitely just host based). If we add something like the fast round or accurate mod that'd need doing though

File: inst/library.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+// These exist to support the model on the gpu, as in C++14 std::min
+// and std::max are constexpr and error without --expt-relaxed-constexpr
+template <typename T>
+HOSTDEVICE T odin_min(T x, T y) {
+  return x < y ? x : y;
+}
+
+template <typename T>
+HOSTDEVICE T odin_max(T x, T y) {
+  return x < y ? x : y;
```
Comment (johnlees) [inline inst/library.hpp:19]:
```suggestion
  return x < y ? y : x;
```

File: tests/testthat/test-sexp.R [hunk 3]
```diff
@@ -65,14 +65,24 @@ test_that("2-arg round is not supported", {
...
+
+  expect_equal(
+    generate_dust_sexp(list("min", "a"), NULL, NULL, NULL, TRUE),
+    "a")
+  expect_equal(
+    generate_dust_sexp(list("min", "a", "b"), NULL, NULL, NULL, TRUE),
+    "odin_min(a, b)")
+  expect_equal(
+    generate_dust_sexp(list("min", "a", "b", "c"), NULL, NULL, NULL, TRUE),
+    "odin_min(a, odin_min(b, c))")
```
Comment (johnlees) [inline tests/testthat/test-sexp.R:85]: Haven't tested max?

**Review by johnlees — CHANGES_REQUESTED** [review]

**Review by johnlees — APPROVED** [review]

**Pull request merged**